### PR TITLE
[DO NOT MERGE][Object Store HA][ProtoType] Object Store HA 

### PR DESCRIPTION
**Pull request by Catch-Bull**
## Why are these changes needed?

This is a prototype test of Object Store HA. REP: https://github.com/ray-project/enhancements/pull/10
The object which supports HA we called `HA_Object`:
- To reduce workload, we use spill/restore objects to implement dump/load checkpoints. In the final PR, we still tend to implement dump/load checkpoints alone on the basis of using the existing code as much as possible.
- Although we call `PinObjects` when we seal `HA_Object`, `HA_Object` doesn't have the primary copy, it is only so that it can be spilled. When the global owner restarts and the global references are released, there is no good idea on how to delete the primary copy. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
The current prototype is:

* An HA object is spilled on creation. Since it's spilled to remote storage, it doesn't have a primary copy. All copies in plasma are secondary copies.
* We assume that we don't GC data on remote storage, even if an HA object is out-of-scope, so we don't care about reference counting either. Global owners are only responsible for location recording to avoid unnecessary restoring from remote storage. It doesn't need to manage references.

**Comment by stephanie-wang** [comment]
Will try to review this soon, but we're a bit busy at the moment with the 2.0 release.

**Comment by ericl** [comment]
This is very exciting. Also looking forwards to reviewing after 2.0.

**Comment by stephanie-wang** [comment]
Thanks for the prototype! A few things:
- It seems like it would simplify the code a lot if we make actor_id and checkpoint_url optional fields of the object's ownership info
- I found it a bit hard to follow all the pieces of code since there is a lot of boilerplate code as well. Can you summarize the major changes that were made to each component?
- How do we delete HA objects / GC the metadata? It seems like there could be a situation where we end up losing all the language-level copies of the ObjectRef, so no one can call `free`, but the system still thinks that there is such an object. Can we rethink the API to support this case? 
- How will it work for objects created by .remote()?

**Comment by Catch-Bull** [comment]
> Thanks for the prototype! A few things:
> 
> * It seems like it would simplify the code a lot if we make actor_id and checkpoint_url optional fields of the object's ownership info
> * I found it a bit hard to follow all the pieces of code since there is a lot of boilerplate code as well. Can you summarize the major changes that were made to each component?
> * How do we delete HA objects / GC the metadata? It seems like there could be a situation where we end up losing all the language-level copies of the ObjectRef, so no one can call `free`, but the system still thinks that there is such an object. Can we rethink the API to support this case?
> * How will it work for objects created by .remote()?

- Thanks for the reminder, I will try it on the next PR.
- no problem, I will also implement HA-returned-object in the prototype later, and I will summarize it together later.
- Since HA-Object no longer has a primary copy, HA-Object's metadata no longer has the reference count, but only contains location information. It is indeed possible that HA-Object's reference does not exist, but its data in plasma is not deleted, and the location still exists, but I think this "leak" is ok.
- I will implement HA-returned-object in this prototype later.
